src/browser/starter.js

```diff
@@ -175,54 +175,68 @@ function V86Starter(options)
         v86_bin_fallback = "build/" + v86_bin_fallback;
     }
 
-    v86util.load_file(v86_bin, {
-        done: bytes =>
-        {
-            WebAssembly
-                .instantiate(bytes, { "env": wasm_shared_funcs })
-                .then(({ instance }) => {
-                    const imports = wasm_shared_funcs;
-                    const exports = instance["exports"];
-                    wasm_memory = exports.memory;
-                    exports["rust_init"]();
-
-                    const emulator = this.v86 = new v86(this.emulator_bus, { exports, wasm_table });
-                    cpu = emulator.cpu;
-
-                    this.continue_init(emulator, options);
-                }, err => {
-                    v86util.load_file(v86_bin_fallback, {
-                        done: bytes => {
-                            WebAssembly
-                                .instantiate(bytes, { "env": wasm_shared_funcs })
-                                .then(({ instance }) => {
-                                    const imports = wasm_shared_funcs;
-                                    const exports = instance["exports"];
-                                    wasm_memory = exports.memory;
-                                    exports["rust_init"]();
-
-                                    const emulator = this.v86 = new v86(this.emulator_bus, { exports, wasm_table });
-                                    cpu = emulator.cpu;
-
-                                    this.continue_init(emulator, options);
-                                });
-                        },
+    if (options["wasm_fn"]) {
+        options["wasm_fn"]({ "env": wasm_shared_funcs })
+            .then((exports) => {
+                const imports = wasm_shared_funcs;
+                wasm_memory = exports.memory;
+                exports["rust_init"]();
+
+                const emulator = this.v86 = new v86(this.emulator_bus, { exports, wasm_table });
+                cpu = emulator.cpu;
+
+                this.continue_init(emulator, options);
```

Comment:
Could you refactor this into a function? (or, even better, refactor it to use async/await)

Reply:
Unfortunately since it's a constructor I can't make it async, is it ok if I make V86Starter call async stuff without returning a promise? Users would need to listen to emulator-ready

Reply:
Updated it btw